File: ios/DownloadManagerKit/Sources/DownloadManagerKit/DownloadStore.swift
//
//  DownloadStore.swift
//  DownloadManagerKit
//

import CoreFoundation
import Foundation
import os.log

private let currentSchemaVersion: UInt32 = 1

private enum StoreDecodingError: LocalizedError {
   case malformedEnvelope
   case unsupportedVersion(UInt32)
   case invalidRecords

   var errorDescription: String? {
      switch self {
      case .malformedEnvelope:
         return "Malformed store envelope"
      case .unsupportedVersion(let version):
         return "Unsupported store version: \(version) (expected \(currentSchemaVersion))"
      case .invalidRecords:
         return "Invalid store records"
      }
   }
}

/// Reject Boolean and floating-point versions before JSONDecoder converts
/// whole-valued decimals and exponents to UInt32. StoreDocument checks the range,
/// supported version, and record payload after this number-type check.
private func validateVersionToken(in data: Data) throws {
   guard let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
         let version = root["version"] as? NSNumber,
         CFGetTypeID(version) != CFBooleanGetTypeID(),
         !CFNumberIsFloatType(version)
   else { throw StoreDecodingError.malformedEnvelope }
}

/// The current record type remains the v1 payload until a real migration is needed.
private struct StoreDocument: Codable {
   let version: UInt32
   let downloads: [DownloadRecord]

   private enum CodingKeys: String, CodingKey {
      case version, downloads
   }

   init(downloads: [DownloadRecord]) {
      self.version = currentSchemaVersion
      self.downloads = downloads
   }

   init(from decoder: Decoder) throws {
      let container: KeyedDecodingContainer<CodingKeys>
      do {
         container = try decoder.container(keyedBy: CodingKeys.self)
         version = try container.decode(UInt32.self, forKey: .version)
         // Check the array shape without decoding records from a future schema.
         _ = try container.nestedUnkeyedContainer(forKey: .downloads)
      } catch {
         throw StoreDecodingError.malformedEnvelope
      }

      guard version == currentSchemaVersion else {
         throw StoreDecodingError.unsupportedVersion(version)
      }
      do {
         downloads = try container.decode([DownloadRecord].self, forKey: .downloads)
      } catch {
         throw StoreDecodingError.invalidRecords
      }
   }
}

/// Thread-safe store for versioned download records, persisted to an atomically
/// written JSON file.
actor DownloadStore {
   private var downloads: [DownloadRecord]
   private let savePath: URL

   /// - Parameter savePath: Where the store is persisted. Injectable so tests can
   ///   work against a temporary file rather than the app's Application Support
   ///   directory.
   ///   Defaults through [`StoreLocation`], which is what a configured directory
   ///   reaches this class by — the path has to be known here, since the file is read
   ///   below rather than on first use.
   init(savePath: URL = StoreLocation.savePath()) {
      self.savePath = savePath
      self.downloads = DownloadStore.load(from: savePath)
   }

   func list() -> [DownloadRecord] { downloads }
   
   func findByPath(_ path: String) -> DownloadRecord? {
      downloads.first(where: { $0.path == path })
   }
   
   func append(_ item: DownloadRecord) {
      downloads.append(item)
      save()
   }
   
   func update(_ item: DownloadRecord, persist: Bool = true) {
      if let index = downloads.firstIndex(where: { $0.path == item.path }) {
         downloads[index] = item
      }
      if persist {
         save()
      }
   }
   
   /// Records a newly-learned total in a single actor hop, returning the updated
   /// record, or nil when the total was already known.
   ///
   /// Progress callbacks arrive as unordered tasks, so a compare-and-set composed
   /// from a separate `findByPath` and `update` lets several callbacks each read an
   /// unknown total and each act on it. Deliberately does not persist: pause,
   /// cancel and completion all write the record, and this runs on the hottest
   /// callback in the system.
   func setTotalIfChanged(path: String, total: UInt64) -> DownloadRecord? {
      guard let index = downloads.firstIndex(where: { $0.path == path }),
            downloads[index].totalBytes != total else {
         return nil
      }

      downloads[index].setBytes(received: downloads[index].receivedBytes, total: total)

      return downloads[index]
   }

   /// Applies `body` in one actor hop and returns the stored record, or nil when no
   /// record has that path. Composed from `findByPath` and `update` it would suspend
   /// between read and write and lose concurrent changes; `body` is synchronous for
   /// the same reason.
   func mutate(path: String, persist: Bool, _ body: @Sendable (inout DownloadRecord) -> Void) -> DownloadRecord? {
      guard let index = downloads.firstIndex(where: { $0.path == path }) else {
         return nil
      }

      body(&downloads[index])

      if persist {
         save()
      }

      return downloads[index]
   }

   /// Applies several record updates with a single write.
   ///
   /// Reconciliation can revert many records at once; one write per record would
   /// re-encode and rewrite the whole file that many times. Unknown paths are
   /// ignored.
   func update(_ records: [DownloadRecord]) {
      guard !records.isEmpty else { return }

      for record in records {
         if let index = downloads.firstIndex(where: { $0.path == record.path }) {
            downloads[index] = record
         }
      }

      save()
   }

   func remove(_ item: DownloadRecord) {
      if let index = downloads.firstIndex(where: { $0.path == item.path }) {
         downloads.remove(at: index)
      }
      save()
   }
   
   /// Decodes the persisted store.
   ///
   /// Rejects the whole document if any record is malformed. Preserving an
   /// unreadable file before continuing empty is tracked separately in #64.
   static func load(from savePath: URL) -> [DownloadRecord] {
      do {
         let data = try Data(contentsOf: savePath)
         return try decodeRecords(from: data)
      } catch {
         os_log(.error, log: Log.downloadStore, "Failed to load download store: %{public}@", error.localizedDescription)
         return []
      }
   }

   /// Separate decoding from the existing log-and-continue load path so tests can
   /// distinguish invalid envelopes, unsupported versions, and invalid records.
   static func decodeRecords(from data: Data) throws -> [DownloadRecord] {
      do {
         try validateVersionToken(in: data)
         return try JSONDecoder().decode(StoreDocument.self, from: data).downloads
      } catch let error as StoreDecodingError {
         throw error
      } catch {
         // Raw decoder errors can expose document contents in the store logger.
         throw StoreDecodingError.malformedEnvelope
      }
   }

   private func save() {
      let encoder = JSONEncoder()
      do {
         let data = try encoder.encode(StoreDocument(downloads: downloads))

         // `write` does not create intermediate directories. `StoreLocation.set` already
         // created a configured directory, and is where an unusable one fails loudly;
         // this covers the rest — the default path, which never goes through `set`, and
         // a directory removed while the app runs. A failure here reaches only the
         // `catch`, so it must not be the first line of defence.
         // Desktop creates it in `save_inner`; Android gets it from `AtomicFile`.
         try FileManager.default.createDirectory(
            at: savePath.deletingLastPathComponent(),
            withIntermediateDirectories: true
         )

         try data.write(to: savePath, options: .atomic)
      } catch {
         os_log(.error, log: Log.downloadStore, "Failed to save download store: %{public}@", error.localizedDescription)
      }
   }
}

File: android/lib/src/main/java/org/silvermine/downloadmanager/DownloadStore.kt
package org.silvermine.downloadmanager

import android.util.AtomicFile
import android.util.Log
import kotlinx.serialization.Serializable
import kotlinx.serialization.SerializationException
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.decodeFromJsonElement
import kotlinx.serialization.json.longOrNull
import java.io.File

/** The envelope fields have no defaults so both are always required and written. */
@Serializable
private data class StoreDocument(val version: Int, val downloads: List<DownloadRecord>)

/**
 * Thread-safe store for download records backed by an atomic JSON file.
 *
 * All public methods are synchronized to ensure consistency when accessed
 * from multiple threads (e.g. WorkManager workers and the main thread).
 * Mirrors the iOS DownloadStore actor pattern.
 *
 * @param directory The directory holding [STORE_FILENAME]. Taken rather than derived
 *    from a Context so the location is configurable; the caller resolves the default.
 *    Read here and nowhere else, which is why it cannot be changed after construction.
 */
internal class DownloadStore(directory: File) {
   private val file = AtomicFile(storeFile(directory))
   private val downloads = mutableMapOf<String, DownloadRecord>()

   init {
      load()
   }

   @Synchronized
   fun list(): List<DownloadRecord> = downloads.values.toList()

   @Synchronized
   fun findByPath(path: String): DownloadRecord? = downloads[path]

   @Synchronized
   fun append(record: DownloadRecord) {
      downloads[record.path] = record
      save()
   }

   @Synchronized
   fun update(record: DownloadRecord, persist: Boolean = true) {
      if (downloads.containsKey(record.path)) {
         downloads[record.path] = record
      }
      if (persist) {
         save()
      }
   }

   /**
    * Applies several record updates with a single write.
    *
    * Reconciliation can revert many records at once; one write per record would
    * rewrite the whole file that many times.
    *
    * @param records The records to update. Unknown paths are ignored.
    */
   @Synchronized
   fun update(records: List<DownloadRecord>) {
      if (records.isEmpty()) {
         return
      }

      for (record in records) {
         if (downloads.containsKey(record.path)) {
            downloads[record.path] = record
         }
      }

      save()
   }

   @Synchronized
   fun remove(record: DownloadRecord) {
      downloads.remove(record.path)
      save()
   }

   private fun load() {
      try {
         val records = decodeRecords(String(file.readFully()))
         downloads.clear()
         for (record in records) {
            downloads[record.path] = record
         }
      } catch (e: Exception) {
         Log.e(TAG, "Failed to load download store: ${e.message}")
      }
   }

   private fun save() {
      val bytes = encodeRecords(downloads.values.toList()).toByteArray()
      val stream = file.startWrite()
      try {
         stream.write(bytes)
         file.finishWrite(stream)
      } catch (e: Exception) {
         file.failWrite(stream)
         Log.e(TAG, "Failed to save download store: ${e.message}")
      }
   }

   companion object {
      private const val TAG = "DownloadStore"
      private const val STORE_FILENAME = "downloads.json"
      private const val CURRENT_SCHEMA_VERSION = 1

      /**
       * Resolves the store file inside a directory.
       *
       * @param directory The directory holding the store.
       * @return The store file.
       */
      internal fun storeFile(directory: File): File = File(directory, STORE_FILENAME)

      private val json = Json { ignoreUnknownKeys = true }

      /**
       * Decodes persisted records.
       *
       * Rejects the whole document if any record is malformed. Preserving an
       * unreadable file before continuing empty is tracked separately in #64.
       *
       * Extracted from [load] so it can be unit-tested, and deliberately free of
       * logging to keep it so: `android.util.Log` is a throwing stub off-device.
       *
       * @param text The persisted store's contents.
       * @return The decoded records.
       */
      internal fun decodeRecords(text: String): List<DownloadRecord> {
         val root = try {
            json.parseToJsonElement(text)
         } catch (_: SerializationException) {
            // Decoder messages can contain private values from the input.
            throw SerializationException("Malformed store envelope")
         }
         val document = root as? JsonObject
            ?: throw SerializationException("Malformed store envelope")
         val versionField = document["version"] as? JsonPrimitive
         val version = versionField?.takeUnless { it.isString }?.longOrNull
         val records = document["downloads"] as? JsonArray
         if (version == null || version < 0 || records == null) {
            throw SerializationException("Malformed store envelope")
         }
         if (version != CURRENT_SCHEMA_VERSION.toLong()) {
            throw SerializationException("Unsupported store version: $version (expected $CURRENT_SCHEMA_VERSION)")
         }

         return try {
            json.decodeFromJsonElement<List<DownloadRecord>>(records)
         } catch (_: SerializationException) {
            throw SerializationException("Invalid store records")
         }
      }

      /**
       * Encodes records for persistence.
       *
       * @param records The records to encode.
       * @return The JSON text to persist.
       */
      internal fun encodeRecords(records: List<DownloadRecord>): String =
         json.encodeToString(StoreDocument(CURRENT_SCHEMA_VERSION, records))
   }
}
